### PR TITLE
make dyanmo work with _LazyGraphModule.lazy_forward

### PR DESCRIPTION
**Pull request by shunting314**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #121259

Fix https://github.com/pytorch/pytorch/issues/121198 .

We previously already trigger the real recompilation for LazyGraphModule when it runs thru dynamo context. But people may pass in LazyGraphModule._lazy_forward rather than the LazyGraphModule instance itself. This PR handles that.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

### CODE REVIEW

File: torch/_dynamo/eval_frame.py [hunk 1]
```diff
@@ -356,21 +356,31 @@ def get_compiler_config():
         fn = innermost_fn(fn)
 
         # add context containing GraphModule to any GraphModule forward functions
-        if isinstance(fn, torch.fx.GraphModule):
```
Comment (williamwen42) [inline torch/_dynamo/eval_frame.py:359]: Why are we removing this case? It seems like `test_preserve_fx_node_metadata` is failing as a result.

**Review by williamwen42 — CHANGES_REQUESTED** [review]

**Review by jansel — CHANGES_REQUESTED** [review]
Tests are failing

File: torch/_dynamo/eval_frame.py (hunk 1 above)
Comment (shunting314) [reply torch/_dynamo/eval_frame.py:359]: Didn't realize some code is ran for non _LazyGraphModule in this code block. Fixing it now.

**Review by williamwen42 — APPROVED** [review]

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**